pkg/controller/direct/compute/refs.go

```diff
@@ -0,0 +1,475 @@
+	}
+
+	key := types.NamespacedName{
+		Namespace: ref.Namespace,
+		Name:      ref.Name,
+	}
+	if key.Namespace == "" {
+		key.Namespace = src.GetNamespace()
+	}
+
+	computeTargetGrpcProxy, err := resolveResourceName(ctx, reader, key, schema.GroupVersionKind{
+		Group:   "compute.cnrm.cloud.google.com",
+		Version: "v1beta1",
+		Kind:    "ComputeTargetGrpcProxy",
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// targetField: self_link
+	// See compute servicemappings for details
+	return &refs.ComputeTargetGrpcProxyRef{
+		External: computeTargetGrpcProxy.GetSelfLink()}, nil
+}
+
+func ResolveComputeTargetHTTPProxy(ctx context.Context, reader client.Reader, src client.Object, ref *refs.ComputeTargetHTTPProxyRef) (*refs.ComputeTargetHTTPProxyRef, error) {
+	if ref == nil {
+		return nil, nil
+	}
+
+	if ref.External != "" {
+		if ref.Name != "" {
+			return nil, fmt.Errorf("cannot specify both name and external on reference")
+		}
+		return ref, nil
+	}
+
+	if ref.Name == "" {
+		return nil, fmt.Errorf("must specify either name or external on reference")
+	}
+
+	key := types.NamespacedName{
+		Namespace: ref.Namespace,
+		Name:      ref.Name,
+	}
+	if key.Namespace == "" {
+		key.Namespace = src.GetNamespace()
+	}
+
+	computeTargetHTTPProxy, err := resolveResourceName(ctx, reader, key, schema.GroupVersionKind{
+		Group:   "compute.cnrm.cloud.google.com",
+		Version: "v1beta1",
+		Kind:    "ComputeTargetHTTPProxy",
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// targetField: self_link
+	// See compute servicemappings for details
+	return &refs.ComputeTargetHTTPProxyRef{
+		External: computeTargetHTTPProxy.GetSelfLink()}, nil
+}
+
+func ResolveComputeTargetHTTPSProxy(ctx context.Context, reader client.Reader, src client.Object, ref *refs.ComputeTargetHTTPSProxyRef) (*refs.ComputeTargetHTTPSProxyRef, error) {
+	if ref == nil {
+		return nil, nil
+	}
+
+	if ref.External != "" {
+		if ref.Name != "" {
+			return nil, fmt.Errorf("cannot specify both name and external on reference")
+		}
+		return ref, nil
+	}
+
+	if ref.Name == "" {
+		return nil, fmt.Errorf("must specify either name or external on reference")
+	}
+
+	key := types.NamespacedName{
+		Namespace: ref.Namespace,
+		Name:      ref.Name,
+	}
+	if key.Namespace == "" {
+		key.Namespace = src.GetNamespace()
+	}
+
+	computeTargetHTTPSProxy, err := resolveResourceName(ctx, reader, key, schema.GroupVersionKind{
+		Group:   "compute.cnrm.cloud.google.com",
+		Version: "v1beta1",
+		Kind:    "ComputeTargetHTTPSProxy",
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// targetField: self_link
+	// See compute servicemappings for details
+	return &refs.ComputeTargetHTTPSProxyRef{
+		External: computeTargetHTTPSProxy.GetSelfLink()}, nil
+}
+
+func ResolveComputeTargetSSLProxy(ctx context.Context, reader client.Reader, src client.Object, ref *refs.ComputeTargetSSLProxyRef) (*refs.ComputeTargetSSLProxyRef, error) {
+	if ref == nil {
```

Comment:
As we build more of these, we should look at whether it's possible to have more helper functions for the common cases (similar to what you've done with resolveResourceName) e.g.

```
func ResolveComputeTargetSSLProxy(ctx context.Context, reader client.Reader, src client.Object, ref *refs.ComputeTargetSSLProxyRef) (*refs.ComputeTargetSSLProxyRef, error) {
  external, err := resolveNormalRef(ctx, reader, src, ref.Name, ref.Namespace, ref.External, schema.GroupVersionKind{...})
  if err != nil { ... }
  return &refs.ComputeTargetSSLProxyRef{ External: external }
}
````